proposals/p1885.md

```diff
@@ -0,0 +1,779 @@
+  impl as IterateBidirectional ... { ... }
+}
+var container: MyContainer = MyContainer.Create(...);
+for (v: container.(Iterate.ElementType) in container.Reversed()) {...}
+```
+
+#### Interoperability with C++ types and containers
+
+The cursor approach deviates from the C++ iterator approach, requiring some
+adaptation.
+
+Given the current state of the design, C++ interoperability is yet to be
+defined, and suggestions have been highlighted in the
+[Future work](#future-work) section.
+
+## Future work
+
+### `for` with mutable values
+
+Supporting mutable values could be achieved using a second interface that
+provides a proxy or view for the data, exposing an `Iterate` interface with
+`ElementType*`
+
+```carbon
+// Object implementing `MutableIterate` returned by a function
+for (p: T* in my_range.AddrRange()) {}
+```
+
+The example `MutableIterate` interface below acts as a proxy for the collection,
+and implements our `Iterate` interface.
+
+```carbon
+interface MutableIterate {
+  impl as Iterate;
+  alias ElementType = Iterate.ElementType;
+  let ProxyType:! Iterate where .ElementType = ElementType*
+      and .CursorType is ImplicitAs((Self as Iterate).CursorType);
+  fn AddrRange[addr self: Self*]() -> ProxyType;
+}
+
+class MyIntContainer {
+  //...
+  impl as Iterate ...
+  external impl as MutateIterate where .ProxyType = Self {
+    ...
+  }
+}
+```
+
+### Speculative mixin usage
+
+Mixins are currently in early design stages. This section highlights possible
+uses speculating on the final design. Some may include:
+
+-   Improved semantics for views compared to getter methods: no direct side
+    effect from using the view
+-   Facilitate code reuse, compared to reimplementing an interface
+-   Direct access to `self`, limiting needs for pointers and address resolution
+
+#### Mutable values
+
+Named mixins, used as programmable members, could expose a view with mutable
+values:
+
+```carbon
+// Using a named mixin to expose mutable values
+for (p: my_range.(Iterate.ElementType)* in my_range.mutable) {}
+```
+
+Used in this way, this clarifies that this a view of the data, like an
+attribute, with no direct side effects, when compared to a method.
+
+#### Alternative views
+
+Similarly, mixins could allow supporting other views into the container.
+
+For example, maps could expose a view with "key" and "value" only, in addition
+to the default (key, value) tuple.
+
+```carbon
+// Using a named mixin to expose mutable values
+for (k: my_dict.(Iterate.CursorType) in my_dict.keys) {}
+```
+
+And "reversed" view could also be used for reverse iteration:
+
+```carbon
+// Reversing
+for (v: my_dict.(Iterate.ElementType) in my_dict.reversed) {}
+```
+
+### Large elements and Optional
+
+Large elements would have a negative impact if the `Optional` cannot leverage
+unused bit patterns within the type, nor present a view instead of a copy.
+
+Such opportunities for `Optional` exist, and could be addressed in the future.
```

Comment:
I'm skeptical that there are opportunities for `Optional(T)` to present a view of the `T` object instead of a copy. Currently, Carbon has no way of expressing that, and there is substantial resistance to adding a way to do so, because a view of an arbitrary type is basically a reference. As far as I can tell, if we want to solve this problem without introducing references, we have to do it as part of the design of `for`, not the design of `Optional`. That way we can define the reference-like behavior as part of the hidden machinery that `for` rewrites to, without giving user code access to the full power of reference types. See #2274 for an example of something similar, where we enabled a particular kind of reference-like behavior by defining special rules for how `[]` is rewritten to a method call.

In many ways inversion of control would be an ideal solution here; it's really unfortunate that it's apparently so hard to optimize.

Reply:
If I understand that correctly, you'd suggest having a similar rewrite that bypasses `Optional`s entirely. Which means needing to expose an `Optional`-less API (but then losing the advantage of having a single `Next` method), or short-circuiting the `return` before the Optional is created (is that possible? probably if we can invert control). I'm not sure the same situation applies exactly, feel free to correct me.

I was skeptical too, @zygoloid seemed somewhat confident about it in a previous discussion. But I haven't done a good job at capturing here what he had in mind. @zygoloid would you mind sharing again what you had in mind if you recall?

Reply:
We discussed this a bit at the weekly meeting today, and a couple of other options came up:
- We could let the container choose whether it exposes values or pointers to values, and maybe provide a dereferencing adapter that turns a range of pointers into a range of values. The nice thing is this doesn't involve any technical changes to the design, just a shift in expectations about how containers will typically implement `Iterate`. However, that seems pretty awkward from a user experience point of view, and I think it would mean that any given container type could support efficient iteration, or iteration over container r-values, but not both, so I think this only makes sense as a stopgap while we investigate a longer-term solution like control inversion.
- We could allow the binding in the `for` loop be an `addr` pattern, which would be implemented by calling into a separate interface on the container. This puts the choice completely in the hands of the `for` loop author, so it wouldn't let us match the ergonomics of `let` (where the compiler can choose to copy or alias the value depending on what's more efficient), but it at least gives us parity with C++ (where the author has to choose whether the iteration variable is a reference).

Reply:
Thank you, updated to reflect the options suggested here, and noted that the preference today seems to be with #1, to get to a first step before revising.